text/001x-battery_improvements.md

```diff
@@ -0,0 +1,283 @@
+- Provide a cleaner message and design for both implementers and consumers.
+- Align with DroneCan battery messages to ease transport and data conversion across different transports.
+
+The [BATTERY_STATUS](https://mavlink.io/en/messages/common.html#BATTERY_STATUS) has two arrays that can be used to report individual cell voltages (up to 14), or a single cumulative voltage, and which cannot be truncated.
+The vast majority of consumers are only interested in the total battery voltage, and either sum the battery cells or get the cumulative voltage.
+By separating the cell voltage reporting into a separate message the new battery status can be much smaller, and the cell voltages need only be sent if the consumer is actually interested.
+
+# Detailed Design
+
+There are three parts to the design:
+1. A more efficient status message
+2. A scalable battery voltage message.
+3. Mechanisms that allow the new messages to seamlessly coexist with the old one along with eventual deprecation of the older message.
+
+## BATTERY_STATUS_V2
+
+The proposed message is:
+
+```xml
+    <message id="???" name="BATTERY_STATUS_V2">
+      <description>Battery dynamic information.
+        This should be streamed (nominally at 1Hz).
+        Static/invariant battery information is sent in SMART_BATTERY_INFO.
+
+        The full-battery charge can be inferred from current_remaining and percent_remaining if both values are supplied.
+        The `current_remaining` field should only be sent if it is guaranteed to be near-accurate.
+        Power monitors do not normally send the `current_remaining` field as it can only be accurate if the battery is fully charged when the drone is turned on.
+        (A GCS can use `current_remaining` being invalid as a trigger to notify the user to fully charge the battery before flight).
+      </description>
+      <field type="uint8_t" name="id" instance="true">Battery ID</field>
+      <field type="int16_t" name="temperature" units="cdegC" invalid="INT16_MAX">Temperature of the whole battery pack (not internal electronics). INT16_MAX field not provided.</field>
+      <field type="uint32_t" name="voltage" units="mV" invalid="UINT32_MAX">Battery voltage (total). UINT32_MAX: field not provided.</field>
+      <field type="uint32_t" name="current" units="mA" invalid="UINT32_MAX">Battery current (through all cells/loads). UINT32_MAX: field not provided.</field>
+      <field type="uint32_t" name="current_consumed" units="mAh" invalid="UINT32_MAX">Consumed charge (since vehicle powered on). UINT32_MAX: field not provided. Note: For power modules the expectation is that batteries are fully charged before turning on the vehicle.</field>
+      <field type="uint32_t" name="current_remaining" units="mAh" invalid="UINT32_MAX">Remaining charge (until empty). UINT32_MAX: field not provided. Note: Power monitors should not set this value.</field>
```

Comment:
@hamishwillee Something similar to the idea was brought up by @hendjoshsr71 for the new DroneCAN messages. It's 4 bytes tho so for something so simple a status flag might make more sense. Idk though how tight we want it to be, we did just add another 4 bytes for 2 additional temperatures. 
https://github.com/dronecan/DSDL/pull/7/commits/2e2f070c4d131d8ae1fae7155a021544c89c4203#diff-06d183633540443cc294a31d649e99bcbe708700160dbb421e0c6204dfadbf9bR9-R10

And I am looking at it now and thinking we really don't need `capacity_remaining` since we would know full and consumed, as long as we redefined consumed to mean total consumed for the pack. This would be coloumb counting with an analog power monitor (consumed since boot) and it would be the fuel gauge for smart batteries (consumed since full)
https://github.com/dronecan/DSDL/pull/7/files#diff-a276641f2405349cc2307f345b18ec6e35498c405f3849842f22840603a90a4aR11

Reply:
Thanks @dakejahl :-). I commented on the post in dronecan: https://github.com/dronecan/DSDL/pull/7#issuecomment-1144239330

The summary is for MAVLink I think we really need to have capacity remaining. 

But what should the status bit be? I lean towards something that tells us what the measurements are relative to:

- MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL : If set , `current_consumed` and `current_remaining` are relative to battery full. If unset, `current_consumed` is relative to drone power-on, and `current_remaining` is an estimate based on the _assumption_ that battery was full on power on. This bit should be set by smart batteries and unset for power modules. If unset a GCS should always advise users to charge the battery on power on.

Reply:
I'm cool with it, that works! But if we always provide both `consumed` and `remaining` then maybe we don't need `percent_remaining`

Reply:
also this is nitpicky but I think we should call it `capacity_consumed` and `capacity_remaining` since mAh are units of capacity

Reply:
Hurray. We're crossing over chats a bit :-)

> also this is nitpicky but I think we should call it capacity_consumed and capacity_remaining since mAh are units of capacity

I like nitpicky :-). As long as it is before implementation. I will rename both of those as you have indicated.

> I'm cool with it, that works! But if we always provide both consumed and remaining then maybe we don't need percent_remaining

MAVLink  allows for systems that don't provide one or the other of these values, and I'd like to avoid the case where someone can't supply this because that will give them a divide by zero error (even though the battery is ready to provide a guess). Perhaps I'm over thinking this and absolutely all values should be optional.

UAVCAN does not have to provide it though if it can always provide the others.. I'll make the changes.

Reply:
>PS @dakejahl Do you think it worth separating the status and the fault numbers a bit to allow us to group the faults and insert new items of the various types later if we need to?

Yup I was actually thinking that might be a good idea. First 16 bits for flags, last 16 for faults?

Reply:
Or even break it into 2 uint16_t fields and reduce the text length for each enum string?

Reply:
The fact that you are talking about byte separation or separate fields tells me that maybe you are thinking of separating them "physically" for easier handling of the different types? I was thinking of grouping but not separating them - putting space around each category so they appear together in lists - not actually handling them differently

In the previous message we did have them separately as faults and modes. We merged them to one status because many of the things we had previously were not really either category. The MAVLink benefit of this is that we have one big field that is going to take a while to fill up (if ever). If we split them either logically in the same field or in separate fields we've got a smaller limit on growth in one or the other category.

Looking at this again, I'm not sure that we would want to separate them because almost everything we have now is actually a fault or should be treated like one - e.g. even "charging" is essentially a fault around using the battery to fly.
The only ones that are actually exceptions to this are MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL and 
MAV_BATTERY_STATUS_FLAGS_READY_TO_USE - this is inferred from absence of faults.

So I'd be tempted to move MAV_BATTERY_STATUS_FLAGS_READY_TO_USE and MAV_BATTERY_STATUS_FLAGS_CAPACITY_RELATIVE_TO_FULL literally to the end - 1 and backfill any future things that are really about usage.

Does that makes sense?

Reply:
Done!